File: server/src/ws/handler.ts
/**
 * WebSocket 消息处理器
 * 管理客户端连接、消息路由、定时广播排行榜
 */
import { Server as SocketServer, Socket } from 'socket.io';
import { getActiveRound, processGuess, getRound, startRound, endRound, createRound } from '../services/game.service.js';
import { getLeaderboard, getGuessCount } from '../services/rank.service.js';

// --- 类型 ---
interface ClientMessage {
  type: string;
  data?: Record<string, unknown>;
}

interface ConnectedClient {
  socket: Socket;
  userId: string;
  username: string;
  avatar: string;
  role: 'viewer' | 'host';
}

// --- 状态 ---
const clients = new Map<string, ConnectedClient>();
let leaderboardTimer: ReturnType<typeof setInterval> | null = null;
let roundTimer: ReturnType<typeof setTimeout> | null = null;

/**
 * 初始化 WebSocket 服务
 */
export function setupWebSocket(io: SocketServer): void {
  io.on('connection', (socket: Socket) => {
    console.log(`[WS] 新连接: ${socket.id}`);

    // 客户端注册（带用户信息）
    socket.on('register', (data: { userId: string; username: string; avatar: string; role: string }) => {
      clients.set(socket.id, {
        socket,
        userId: data.userId,
        username: data.username,
        avatar: data.avatar || '',
        role: (data.role as 'viewer' | 'host') || 'viewer',
      });
      console.log(`[WS] 用户注册: ${data.username} (${data.role})`);

      // 如果有活跃轮次，发送当前状态
      const activeRound = getActiveRound();
      if (activeRound) {
        socket.emit('message', {
          type: 'new_round',
          data: {
            roundId: activeRound.id,
            hint: activeRound.hint,
            duration: activeRound.duration,
            startedAt: activeRound.startedAt,
          },
        });
      }
    });

    // 处理消息
    socket.on('message', async (msg: ClientMessage) => {
      const client = clients.get(socket.id);
      if (!client) return;

      try {
        switch (msg.type) {
          case 'guess':
            await handleGuess(io, client, msg.data || {});
            break;
          case 'start_round':
            await handleStartRound(io, client, msg.data || {});
            break;
          case 'end_round':
            await handleEndRound(io, client, msg.data || {});
            break;
          default:
            socket.emit('message', { type: 'error', data: { message: `未知消息类型: ${msg.type}` } });
        }
      } catch (err) {
        console.error('[WS] 处理消息错误:', err);
        socket.emit('message', { type: 'error', data: { message: '服务器内部错误' } });
      }
    });

    // 断开连接
    socket.on('disconnect', () => {
      const client = clients.get(socket.id);
      if (client) {
        console.log(`[WS] 断开: ${client.username}`);
        clients.delete(socket.id);
      }
    });
  });
}

/**
 * 处理猜测消息
 */
async function handleGuess(io: SocketServer, client: ConnectedClient, data: Record<string, unknown>): Promise<void> {
  const roundId = data.roundId as number;
  const text = (data.text as string || '').trim();

  if (!text) {
    client.socket.emit('message', { type: 'error', data: { message: '请输入猜测内容' } });
    return;
  }

  if (text.length > 50) {
    client.socket.emit('message', { type: 'error', data: { message: '输入过长，最多50字' } });
    return;
  }

  const round = getRound(roundId);
  if (!round || round.status !== 'active') {
    client.socket.emit('message', { type: 'error', data: { message: '当前没有进行中的游戏' } });
    return;
  }

  const result = await processGuess(
    roundId,
    client.userId,
    client.username,
    client.avatar,
    text
  );

  // 发送猜测结果给该用户
  client.socket.emit('message', {
    type: 'guess_result',
    data: {
      roundId,
      text,
      similarity: result.similarity,
      isCorrect: result.isCorrect,
      rank: result.rank,
      message: result.message,
    },
  });

  // 广播排行榜更新给所有人
  await broadcastLeaderboard(io, roundId);
}

/**
 * 主播: 开始新一轮
 */
async function handleStartRound(io: SocketServer, client: ConnectedClient, data: Record<string, unknown>): Promise<void> {
  if (client.role !== 'host') {
    client.socket.emit('message', { type: 'error', data: { message: '只有主播可以开始游戏' } });
    return;
  }

  // 检查是否有活跃轮次
  const active = getActiveRound();
  if (active) {
    client.socket.emit('message', { type: 'error', data: { message: '当前已有进行中的游戏，请先结束' } });
    return;
  }

  const round = await createRound({
    wordId: data.wordId as number | undefined,
    category: data.category as string | undefined,
    difficulty: data.difficulty as number | undefined,
    duration: data.duration as number | undefined,
  });

  const startedRound = startRound(round.id);

  // 广播新一轮开始
  broadcast(io, {
    type: 'new_round',
    data: {
      roundId: startedRound.id,
      hint: startedRound.hint,
      duration: startedRound.duration,
      startedAt: startedRound.startedAt,
    },
  });

  // 启动排行榜定时广播（每3秒）
  startLeaderboardBroadcast(io, startedRound.id);

  // 启动倒计时
  roundTimer = setTimeout(async () => {
    await handleAutoEndRound(io, startedRound.id);
  }, startedRound.duration * 1000);

  console.log(`[Game] 轮次 ${startedRound.id} 开始，答案: ${startedRound.word}`);
}

/**
 * 主播: 手动结束轮次
 */
async function handleEndRound(io: SocketServer, client: ConnectedClient, data: Record<string, unknown>): Promise<void> {
  if (client.role !== 'host') {
    client.socket.emit('message', { type: 'error', data: { message: '只有主播可以结束游戏' } });
    return;
  }

  const roundId = data.roundId as number;
  await finishRound(io, roundId);
}

/**
 * 自动结束轮次（倒计时到）
 */
async function handleAutoEndRound(io: SocketServer, roundId: number): Promise<void> {
  const round = getRound(roundId);
  if (!round || round.status !== 'active') return;

  await finishRound(io, roundId);
}

/**
 * 结束轮次并广播结果
 */
async function finishRound(io: SocketServer, roundId: number): Promise<void> {
  // 清除定时器
  stopLeaderboardBroadcast();
  if (roundTimer) {
    clearTimeout(roundTimer);
    roundTimer = null;
  }

  const endedRound = endRound(roundId);
  const leaderboard = await getLeaderboard(roundId);

  // 广播轮次结束
  broadcast(io, {
    type: 'round_end',
    data: {
      roundId: endedRound.id,
      answer: endedRound.word,
      hint: endedRound.hint,
      leaderboard: leaderboard.slice(0, 10), // Top 10
      totalParticipants: leaderboard.length,
    },
  });

  console.log(`[Game] 轮次 ${roundId} 结束，答案: ${endedRound.word}，参与人数: ${leaderboard.length}`);
}

/**
 * 广播排行榜
 */
async function broadcastLeaderboard(io: SocketServer, roundId: number): Promise<void> {
  const leaderboard = await getLeaderboard(roundId);
  broadcast(io, {
    type: 'leaderboard',
    data: {
      roundId,
      entries: leaderboard.slice(0, 50),
      totalParticipants: leaderboard.length,
    },
  });
}

/**
 * 启动排行榜定时广播
 */
function startLeaderboardBroadcast(io: SocketServer, roundId: number): void {
  stopLeaderboardBroadcast();
  leaderboardTimer = setInterval(async () => {
    await broadcastLeaderboard(io, roundId);
  }, 3000); // 每3秒
}

/**
 * 停止排行榜定时广播
 */
function stopLeaderboardBroadcast(): void {
  if (leaderboardTimer) {
    clearInterval(leaderboardTimer);
    leaderboardTimer = null;
  }
}

/**
 * 广播消息给所有已注册客户端
 */
function broadcast(io: SocketServer, message: ClientMessage): void {
  io.emit('message', message);
}

File: game-server/app/ws/handler.py
"""
Socket.IO 事件处理器
管理客户端连接、消息路由、定时广播排行榜
"""
import asyncio
import logging
from typing import Optional

import socketio

from app.services.game_service import (
    get_active_round, process_guess, get_round,
    start_round, end_round, create_round,
)
from app.services.rank_service import get_leaderboard, get_guess_count

logger = logging.getLogger(__name__)

# 已连接客户端
_clients: dict[str, dict] = {}  # sid -> {userId, username, avatar, role}

# 定时器任务
_leaderboard_task: Optional[asyncio.Task] = None
_round_timer_task: Optional[asyncio.Task] = None

# Socket.IO 实例（在 main.py 中设置）
_sio: Optional[socketio.AsyncServer] = None


def set_sio(sio: socketio.AsyncServer) -> None:
    global _sio
    _sio = sio


def _emit(sid: str, event: str, data: dict) -> None:
    """发送消息给单个客户端"""
    if _sio:
        asyncio.create_task(_sio.emit(event, data, room=sid))


def _broadcast(event: str, data: dict) -> None:
    """广播给所有已注册客户端"""
    if _sio:
        for sid in _clients:
            asyncio.create_task(_sio.emit(event, data, room=sid))


async def _broadcast_leaderboard(round_id: int) -> None:
    """广播排行榜"""
    leaderboard = get_leaderboard(round_id)
    _broadcast("message", {
        "type": "leaderboard",
        "data": {
            "roundId": round_id,
            "entries": leaderboard[:50],
            "totalParticipants": len(leaderboard),
        },
    })


async def _leaderboard_loop(round_id: int) -> None:
    """定时广播排行榜（每3秒）"""
    try:
        while True:
            await asyncio.sleep(3)
            await _broadcast_leaderboard(round_id)
    except asyncio.CancelledError:
        pass


async def _auto_end_round(round_id: int, duration: int) -> None:
    """自动结束轮次"""
    try:
        await asyncio.sleep(duration)
        await _finish_round(round_id)
    except asyncio.CancelledError:
        pass


async def _finish_round(round_id: int) -> None:
    """结束轮次并广播"""
    global _leaderboard_task, _round_timer_task

    if _leaderboard_task and not _leaderboard_task.done():
        _leaderboard_task.cancel()
        _leaderboard_task = None

    if _round_timer_task and not _round_timer_task.done():
        _round_timer_task.cancel()
        _round_timer_task = None

    r = get_round(round_id)
    if not r or r.status != "active":
        return

    ended = end_round(round_id)
    leaderboard = get_leaderboard(round_id)

    _broadcast("message", {
        "type": "round_end",
        "data": {
            "roundId": ended.id,
            "answer": ended.word,
            "hint": ended.hint,
            "leaderboard": leaderboard[:10],
            "totalParticipants": len(leaderboard),
        },
    })
    logger.info(f"轮次 {round_id} 结束，答案: {ended.word}，参与人数: {len(leaderboard)}")


def register_handlers(sio: socketio.AsyncServer) -> None:
    """注册 Socket.IO 事件处理器"""
    set_sio(sio)

    @sio.event
    async def connect(sid: str, environ: dict) -> None:
        logger.info(f"新连接: {sid}")

    @sio.event
    async def disconnect(sid: str) -> None:
        client = _clients.pop(sid, None)
        if client:
            logger.info(f"断开: {client.get('username', sid)}")

    @sio.event
    async def register(sid: str, data: dict) -> None:
        _clients[sid] = {
            "userId": data.get("userId", ""),
            "username": data.get("username", ""),
            "avatar": data.get("avatar", ""),
            "role": data.get("role", "viewer"),
        }
        logger.info(f"用户注册: {data.get('username')} ({data.get('role')})")

        # 发送当前活跃轮次
        active = get_active_round()
        if active:
            _emit(sid, "message", {
                "type": "new_round",
                "data": {
                    "roundId": active.id,
                    "hint": active.hint,
                    "duration": active.duration,
                    "startedAt": active.started_at,
                },
            })

    @sio.event
    async def message(sid: str, msg: dict) -> None:
        global _leaderboard_task, _round_timer_task

        client = _clients.get(sid)
        if not client:
            return

        msg_type = msg.get("type", "")
        data = msg.get("data", {})

        try:
            if msg_type == "guess":
                await _handle_guess(sid, client, data)
            elif msg_type == "start_round":
                await _handle_start_round(client, data)
            elif msg_type == "end_round":
                await _handle_end_round(client, data)
            else:
                _emit(sid, "message", {"type": "error", "data": {"message": f"未知消息类型: {msg_type}"}})
        except Exception as e:
            logger.error(f"处理消息错误: {e}", exc_info=True)
            _emit(sid, "message", {"type": "error", "data": {"message": "服务器内部错误"}})

    async def _handle_guess(sid: str, client: dict, data: dict) -> None:
        round_id = data.get("roundId")
        text = (data.get("text") or data.get("guessText") or "").strip()

        if not text:
            _emit(sid, "message", {"type": "error", "data": {"message": "请输入猜测内容"}})
            return
        if len(text) > 50:
            _emit(sid, "message", {"type": "error", "data": {"message": "输入过长，最多50字"}})
            return

        r = get_round(round_id) if round_id else None
        if not r or r.status != "active":
            _emit(sid, "message", {"type": "error", "data": {"message": "当前没有进行中的游戏"}})
            return

        result = await process_guess(
            round_id, client["userId"], client["username"],
            client["avatar"], text,
        )

        _emit(sid, "message", {
            "type": "guess_result",
            "data": {
                "roundId": round_id,
                "text": text,
                "similarity": result["similarity"],
                "isCorrect": result["isCorrect"],
                "rank": result["rank"],
                "message": result["message"],
            },
        })

        await _broadcast_leaderboard(round_id)

    async def _handle_start_round(client: dict, data: dict) -> None:
        global _leaderboard_task, _round_timer_task

        if client["role"] != "host":
            _emit(None, "message", {"type": "error", "data": {"message": "只有主播可以开始游戏"}})
            return

        active = get_active_round()
        if active:
            return

        r = await create_round(
            word_id=data.get("wordId"),
            category=data.get("category"),
            difficulty=data.get("difficulty"),
            duration=data.get("duration"),
        )
        started = start_round(r.id)

        _broadcast("message", {
            "type": "new_round",
            "data": {
                "roundId": started.id,
                "hint": started.hint,
                "duration": started.duration,
                "startedAt": started.started_at,
            },
        })

        # 启动排行榜广播
        if _leaderboard_task and not _leaderboard_task.done():
            _leaderboard_task.cancel()
        _leaderboard_task = asyncio.create_task(_leaderboard_loop(started.id))

        # 启动自动结束倒计时
        if _round_timer_task and not _round_timer_task.done():
            _round_timer_task.cancel()
        _round_timer_task = asyncio.create_task(_auto_end_round(started.id, started.duration))

        logger.info(f"轮次 {started.id} 开始，答案: {started.word}")

    async def _handle_end_round(client: dict, data: dict) -> None:
        if client["role"] != "host":
            return
        round_id = data.get("roundId")
        if round_id:
            await _finish_round(round_id)
